cohort-evaluator-spark/src/main/java/com/ibm/cohort/cql/spark/optimizer/GraphWalkingElmVisitor.java

```diff
@@ -82,41 +82,42 @@ public R visitLibrary(Library elm, C context) {
     
     @Override
     public R visitExpressionRef(ExpressionRef ref, C context) {
-        R result = defaultResult();
+        R result;
+        
         if( ref instanceof FunctionRef ) {
             result = visitFunctionRef((FunctionRef)ref, context);
         } else {
-            if( ! visited.contains(ref) ) {
-                visited.add(ref);
-                
-                Library library = prepareLibraryVisit(getCurrentLibraryIdentifier(), ref.getLibraryName(), context);
-                try {
-                    // The TranslatedLibrary class that the CqlTranslator produces has this already indexed. If we want to 
-                    // use any ELM without relying on the translator, we need to do the lookup ourselves. It is worth
-                    // considering whether or not we build our own indexing helper instead.
-                    ExpressionDef ed = library.getStatements().getDef().stream().filter( def -> def.getName().equals(ref.getName()) ).findAny().get();
-                    result = visitElement(ed,context);
-                } finally {
-                    unprepareLibraryVisit(ref.getLibraryName());
-                }
-            }
+            result = visitRef(ref, context, defaultResult());
         }
+        
         return result;
     }
-    
+
     @Override
     public R visitFunctionRef(FunctionRef ref, C context) {
         R result = super.visitFunctionRef(ref, context);
-        if( ! visited.contains(ref) ) {
+        
+        return visitRef(ref, context, result);
+    }
+    
+    protected R visitRef(ExpressionRef ref, C context, R defaultResult) {
+    	R result = defaultResult;
+    	if( ! visited.contains(ref) ) {
             visited.add(ref);
             
             Library library = prepareLibraryVisit(getCurrentLibraryIdentifier(), ref.getLibraryName(), context);
             try {
                 // The TranslatedLibrary class that the CqlTranslator produces has this already indexed. If we want to 
                 // use any ELM without relying on the translator, we need to do the lookup ourselves. It is worth
                 // considering whether or not we build our own indexing helper instead.
-                ExpressionDef ed = library.getStatements().getDef().stream().filter( def -> def.getName().equals(ref.getName()) ).findAny().get();
-                result = visitElement(ed,context);
+            	Optional<ExpressionDef> optionalEd = library.getStatements().getDef().stream().filter( def -> def.getName().equals(ref.getName()) ).findAny();
```

Comment:
This is the change sonar requested - checking that the optional value is present before calling .get().  When I was in this class I noticed these two methods were almost identical so I consolidated the logic.